language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1395,102 +1445,50 @@ impl IntegerValue {
-            )
-            .with_message(format!(
-                "vector elem layout mismatch, expected {:?}, got {:?}",
-                ty, v
-            ))),
-        }
-    }
-
-    pub fn native_empty(
-        context: &impl NativeContext,
-        ty_args: Vec<Type>,
-        args: VecDeque<Value>,
-    ) -> VMResult<NativeResult> {
-        debug_assert!(ty_args.len() == 1);
-        debug_assert!(args.is_empty());
-
-        let cost = native_gas(context.cost_table(), NativeCostIndex::EMPTY, 1);
-
-        let container = match &ty_args[0] {
-            Type::U8 => Container::U8(vec![]),
-            Type::U64 => Container::U64(vec![]),
-            Type::U128 => Container::U128(vec![]),
-            Type::Bool => Container::Bool(vec![]),
-            Type::Address => Container::Address(vec![]),
-
-            Type::Signer | Type::Vector(_) | Type::Struct(_) | Type::StructInstantiation(_, _) => {
-                Container::General(vec![])
-            }
-
-            Type::Reference(_) | Type::MutableReference(_) | Type::TyParam(_) => {
-                return Err(VMStatus::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                    .with_message(format!("invalid type param for vector: {:?}", &ty_args[0])))
-            }
-        };
-
-        Ok(NativeResult::ok(
-            cost,
-            vec![Value(ValueImpl::new_container(container))],
-        ))
-    }
+pub const INDEX_OUT_OF_BOUNDS: u64 = NFE_VECTOR_ERROR_BASE + 1;
+pub const POP_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 2;
+pub const DESTROY_NON_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 3;
+
+fn check_elem_layout(ty: &Type, v: &Container) -> VMResult<()> {
+    match (ty, v) {
+        (Type::U8, Container::U8(_))
+        | (Type::U64, Container::U64(_))
+        | (Type::U128, Container::U128(_))
+        | (Type::Bool, Container::Bool(_))
+        | (Type::Address, Container::Address(_))
+        | (Type::Signer, Container::General(_))
+        | (Type::Vector(_), Container::General(_))
+        | (Type::Struct(_), Container::General(_)) => Ok(()),
+        (Type::StructInstantiation(_, _), Container::General(_)) => Ok(()),
+
+        (Type::Reference(_), _) | (Type::MutableReference(_), _) | (Type::TyParam(_), _) => {
+            Err(VMStatus::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .with_message(format!("invalid type param for vector: {:?}", ty)))
+        }
+
+        (Type::U8, _)
+        | (Type::U64, _)
+        | (Type::U128, _)
+        | (Type::Bool, _)
+        | (Type::Address, _)
+        | (Type::Signer, _)
+        | (Type::Vector(_), _)
+        | (Type::Struct(_), _)
+        | (Type::StructInstantiation(_, _), _) => Err(VMStatus::new(
+            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+        )
+        .with_message(format!(
+            "vector elem layout mismatch, expected {:?}, got {:?}",
+            ty, v
+        ))),
+    }
+}
+
+impl VectorRef {
```

Comment:
this api is not super consistent but once @vgao1996 reworks the container and the cast to a vector is should all be much more reasonable.
In fact this API may disappear entirely at that point and an implementation only defined in the natives crate

Reply:
Yeah I'll review this while I continue working on the cleanup. Not immediately seeing how this API could disappear with our planned changes but we'll see.